Shell/LineEditor.cpp

```diff
@@ -37,6 +37,60 @@ void LineEditor::append(const String& string)
     m_cursor = m_buffer.size();
 }
 
+void LineEditor::tab_complete_first_token()
+{
+    auto input = String::copy(m_buffer);
+
+    String path = getenv("PATH");
+    if (path.is_empty())
+        return;
+    auto directories = path.split(':');
+
+    String match;
+
+    // Go through the files in PATH.
+    for (const auto& directory : directories) {
+        CDirIterator programs(directory.characters(), CDirIterator::SkipDots);
+        while (programs.has_next()) {
+            String program = programs.next_path();
+
+            // Check that the file is an executable program.
+            struct stat program_status;
+            StringBuilder program_path;
+            program_path.append(directory.characters());
+            program_path.append('/');
+            program_path.append(program.characters());
+            int stat_error = stat(program_path.to_string().characters(), &program_status);
```

Comment:
We should check if the name is a match before we try to `stat()`. That will avoid a ton of file metadata lookups.